akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowStatefulMapAsyncSpec.scala

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.stream.testkit.StreamSpec
+import akka.stream.testkit.scaladsl.TestSink
+
+import java.util.concurrent.ThreadLocalRandom
+import scala.concurrent.Future
+
+class FlowStatefulMapAsyncSpec extends StreamSpec {
+
+  implicit val ec = system.dispatcher
+
+  def randomSleep(): Unit =
+    Thread.sleep(ThreadLocalRandom.current().nextInt(1, 10))
+
+  "statefulMapAsync" must {
+    "worked in happy case and emit final state properly" in {
+
+      val probe = Source(1 to 3)
+        .statefulMapAsync[Int, String](1)(
+          () =>
+            Future {
+              randomSleep()
+              0
+            },
+          (_, e) =>
+            Future {
+              randomSleep()
+              val state = e
+              state -> s"$e element"
+            },
+          s => Future.successful(Some(s"$s state")),
+          _ + _)
+        .runWith(TestSink())
```

Comment:
I think such an operator is a bit problematic since a part of the point with `statefulMap` is to be able to use a mutable state, but for this operator, since the state is intended to be shared across threads it must be immutable, we could document around this but it is highly likely to be missed by users.

I think perhaps it would be better to see `statefulMap().mapAsync()(identity)` where the element and what is needed from the state is passed downstream where it is (safely) turned into a future by mapAsync. This also has more clear behaviour regarding parallelism > 1 IMO.

Reply:
Mostly agree, but even though it is true that you can safely mutate state directly in f when using `statefulMap`, but i'd argue the API itself is designed more like the immutable style, or by some fp jargon state monad style, otherwise why bothering user to explicitly return the next state everytime.

And after investigation on fs2 and ZIO, I find out they both provide the ability to either run effectual action concurrently or accumulating state but not both, which I think they're analog to mapAsync and statefulMapAsync but parallelism is always limited to 1.

I can see how such an easy to misuse operator could easily becomes a maintenance burden while the real use cases are so rare, so could we leave the full API only for implementing other operators and expose to user a variation that has parallelism limited to 1?